spacy/tokens/doc.pyx

```diff
@@ -1924,6 +2042,180 @@ cdef int [:,:] _get_lca_matrix(Doc doc, int start, int end):
     return lca_matrix
 
 
+@cython.boundscheck(False)  # Deactivate bounds checking
+cdef void _set_prefix_lengths(
```

Comment:
This seems like a nice, reusable utility function to look for offsets of UTF-8 characters, that is not specific to prefix lengths. I think the only non-reusable part is that the output buffer size is predetermined. Does it need to be? What is the overhead of getting all offsets and then slicing the result to get the offsets for the prefix? Especially because you can also reuse the output for the suffix char offsets.

Reply:
Another question (I must admit that at this point I do not have a good-enough overview of the PR yet): couldn't you slice the array in the original Python codepoint representation (or `Py_UCS4` in Cython-land) to the prefix/suffix length and then recode that as UTF-8?

Reply:
The second suggestion was what I started off with, and it was unacceptably slow.

As regards the utility function, I think it's important that the typical configuration doesn't call `set_prefix_lengths()` at all (I'm don't know off the top of my head but I think there are only two languages that use it). Most configurations only use `set_suffix_lengths()` and only need the last few characters in each word.

Reply:
> The second suggestion was what I started off with, and it was unacceptably slow.

Do you know why? Slicing `Py_UCS4` arrays in C-land is cheap? Was it the recoding as UTF-8? (Which should also be plenty fast.)

Reply:
I had a [solution](https://github.com/richardpaulhudson/spaCy/blob/7d8258bec851c3b6dabfeef874f16fb1847d63dc) based completely around `Py_UCS4` and [PEP 393](https://peps.python.org/pep-0393/), which seemed the most sensible way to approach the problem because it meant fixed-width characters. However, on several occasions I found that the PEP 393 code didn't behave as documented (to mention just a trivial example, the `kind` variable is supposed to have value `4` for four-byte Unicode characters, but actually has the value `3`). I normally look for alternatives to such interfaces, because I know from past experience that you can never know what other problems might be lurking unnoticed and come to bite you once your code is live.

I think it was you who mentioned (internal communication) that it was easy to recognize the boundaries of UTF-8 characters using bitmaps, and basically once I ditched `Py_UCS4` and went straight to the UTF-8 representation of each string I never looked back. I think the important point is that with a typical configuration you are only ever interested in the last few characters of each token. This means that a solution that only looks at the relevant characters at the end (and, where relevant, at the beginning) of each token will always be faster than one that handles the whole string of each token (or of the whole document, which is how I did it in the version linked above.

Reply:
I have now looked into merging `get_prefix_lengths()` and `get_suffix_lengths()` into a single method `get_affix_lengths()` with the aim of copying information common to the prefixes and the suffixes to avoid calculating it twice. However, because the methods calculate the number of bytes for the first/last _n_ characters for each of a list of values of _n_, it turns out that the only time a value is generated that is common to the prefixes and suffixes is when each contains the length of the whole word, and it does not seem likely that testing for this (relatively rare) condition would lead to an overall speed benefit.